### PR TITLE
Remove active attr typecasting

### PR DESCRIPTION
**Pull request by brianstien**
This removes the active_attr typecasting module and and implements our own.  Instead of typecasting at the read time of an attribute (or attributes), this moves the typecasting to write time.

This is a substantial performance improvement whenever `attributes`, or `respond_to?` is called on an instance, we do not have to typecast every attribute.

I imported the active_attr typecasters and also kept their same dsl so this is backwards compatible.

In the future I think the attribute store should allow attributes to be written without typecasting, so that we can skip that step when getting data from an adapter, which has likely already done the typecasting.  Currently the attribute store is just a dumb hash, but I think that should be improved.

### CODE REVIEW

**Comment by quixoten** [comment]
:+1:

File: lib/active_remote/attributes.rb [hunk 1]
```diff
@@ -1,5 +1,13 @@
 module ActiveRemote
   module Attributes
+    def attributes
```
Comment (liveh2o) [inline lib/active_remote/attributes.rb]: Why is this necessary? Not opposed, just want to understand.

File: lib/active_remote/serializers/protobuf.rb [hunk 2]
```diff
@@ -127,10 +129,7 @@ def typecasted_value
         end
 
         def typecaster
-          @typecaster ||= begin
-            typecaster = TYPECASTER_MAP[field.type_class]
-            typecaster.new if typecaster
-          end
+          @typecaster ||= TYPECASTER_MAP[field.type_class]
```
Comment (liveh2o) [inline lib/active_remote/serializers/protobuf.rb]: 👏

File: lib/active_remote/typecasting.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      return super if value.nil?
+
+      typecaster = _attribute_typecaster(name)
+      return super unless typecaster
+
+      super(name, typecaster.call(value))
+    end
+
+    def _attribute_typecaster(attribute_name)
+      self.class.attributes[attribute_name][:typecaster] || _typecaster_for(attribute_name)
```
Comment (liveh2o) [inline lib/active_remote/typecasting.rb]: Would it make sense to cache this so it's only a hash lookup after the first time it's called?

File: lib/active_remote/typecasting/big_decimal_typecaster.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    class BigDecimalTypecaster
+      def self.call(value)
+        if value.is_a? BigDecimal
+          value
+        elsif value.is_a? Rational
+          value.to_f.to_d
+        elsif value.respond_to? :to_d
+          value.to_d
+        else
+          BigDecimal.new value.to_s
```
Comment (liveh2o) [inline lib/active_remote/typecasting/big_decimal_typecaster.rb]: Parens?

File: lib/active_remote/typecasting/boolean_typecaster.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module ActiveRemote
+  module Typecasting
+    class BooleanTypecaster
+       FALSE_VALUES = ["n", "N", "no", "No", "NO", "false", "False", "FALSE", "off", "Off", "OFF", "f", "F"]
+
+      def self.call(value)
+        case value
+        when *FALSE_VALUES then false
```
Comment (liveh2o) [inline lib/active_remote/typecasting/boolean_typecaster.rb]: I think this is the one @abrandoned said could be more efficient if we check the non-string values first. ActiveModel handles it [a bit differently](https://github.com/rails/rails/blob/52ce6ece8c8f74064bb64e0a0b1ddd83092718e1/activemodel/lib/active_model/type/boolean.rb). Perhaps we could borrow their casting pattern, but add the additional values this supports for compatibility?

File: lib/active_remote/typecasting/date_typecaster.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+module ActiveRemote
+  module Typecasting
+    class DateTypecaster
+      def self.call(value)
+        value.to_date if value.respond_to? :to_date
+      rescue NoMethodError, ArgumentError
```
Comment (liveh2o) [inline lib/active_remote/typecasting/date_typecaster.rb]: What if we just rescued this instead of checking `respond_to?`? I'm also not sure why `ActiveAttr` only provides this on the date typecaster.

File: lib/active_remote/typecasting/object_typecaster.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+module ActiveRemote
+  module Typecasting
+    class ObjectTypecaster
+      def self.call(value)
+        value
```
Comment (liveh2o) [inline lib/active_remote/typecasting/object_typecaster.rb]: 💥

File: lib/active_remote/serializers/protobuf.rb [hunk 8]
```diff
@@ -1,24 +1,26 @@
+require "active_remote/typecasting"
+
 module ActiveRemote
   module Serializers
     module Protobuf
       extend ActiveSupport::Concern
 
       TYPECASTER_MAP = {
```
Comment (fergmastaflex) [inline lib/active_remote/serializers/protobuf.rb]: You're calling this constant `TYPECASTER_MAP`, but the one in typecasting.rb seems to be more of the `TYPECASTING_MAP`. This seems to deal more in setting types specifically on proto fields.

File: lib/active_remote/typecasting/boolean_typecaster.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+        when Numeric, /^\-?[0-9]/ then !value.to_f.zero?
```
Comment (fergmastaflex) [inline lib/active_remote/typecasting/boolean_typecaster.rb]: Do people ever put a "9" to represent `true`?

File: lib/active_remote/typecasting/date_typecaster.rb (hunk 6 above)
Comment (liveh2o) [inline lib/active_remote/typecasting/date_typecaster.rb]: Actually, it's considerably slower:

```
ruby 2.3.0p0 (2015-12-25 revision 53290) [x86_64-darwin15]
                      user     system      total        real
rescue:          24.840000   0.480000  25.320000 ( 25.459208)
condition:        1.400000   0.000000   1.400000 (  1.418840)
rescue/condition 17.742857        Inf        NaN ( 17.943681)
```

Based on http://greyblake.com/blog/2012/09/02/ruby-perfomance-tricks/

File: lib/active_remote/attributes.rb (hunk 1 above)
Comment (liveh2o) [inline lib/active_remote/attributes.rb]: 👍

File: lib/active_remote/typecasting/boolean_typecaster.rb (hunk 9 above)
Comment (liveh2o) [inline lib/active_remote/typecasting/boolean_typecaster.rb]: This regex is checking for strings that are only numbers. Its really a poor way of handling this. I think ActiveModel's [boolean typecaster](https://github.com/rails/rails/blob/52ce6ece8c8f74064bb64e0a0b1ddd83092718e1/activemodel/lib/active_model/type/boolean.rb) is much better.

File: lib/active_remote/typecasting.rb (hunk 3 above)
Comment (brianstien) [inline lib/active_remote/typecasting.rb]: It could be cached, but along with the change to typecasting on writes instead of reads, it is unlikely that a cache will be beneficial since is it rare for attributes to be written more than once during a request/response.

I think future improvements that allow a way to bypass typecasting for data coming from data adapters would simply bypass this codepath entirely for reads.

**Comment by brianstien** [comment]
Regarding the performance of the typecasters, I feel that the best way to improve their performance is to skip them entirely for data coming from the data adapters.  Since the data adapters often already handle typecasting, if the attribute storage mechanism allowed a way to write attributes without typecasting the data adapters could skip this code.

I simply brought over the existing active_attr typecasters in an attempt to make sure we are not introducing breaking changes in this release.

So with that in mind, I would prefer to leave the typecasters alone for now.  I think with a bit of work the typecasters will be optimized out of reads entirely.

File: lib/active_remote/serializers/protobuf.rb (hunk 8 above)
Comment (brianstien) [inline lib/active_remote/serializers/protobuf.rb]: This is the data adapter's typecasting.  The typecasters that were causing performance issues are the user typecasters.  Since I was ripping out the active_attr typecasting system I needed to change this file, since out data adapter typecasters were relying on the active_attr typecasters.

As a side note, most of this is probably not necessary at all since the protobuf library is already handling most of the typecasting.

File: lib/active_remote/typecasting/boolean_typecaster.rb (hunk 5 above)
Comment (brianstien) [inline lib/active_remote/typecasting/boolean_typecaster.rb]: The current state of typecasting leaves a lot to be desired, and certainly needs more work.  As it stands, switching to typecasting at write time instead of read solves the majority of the performance issues.  But there is still a lot more to do, see below.

File: lib/active_remote/serializers/protobuf.rb (hunk 8 above)
Comment (liveh2o) [inline lib/active_remote/serializers/protobuf.rb]: Let's leave this as is for now and revisit it later.

File: lib/active_remote/typecasting.rb (hunk 3 above)
Comment (liveh2o) [inline lib/active_remote/typecasting.rb]: That's fair.

File: lib/active_remote/serializers/protobuf.rb (hunk 8 above)
Comment (fergmastaflex) [inline lib/active_remote/serializers/protobuf.rb]: 👍

**Comment by liveh2o** [comment]
ezgif com-add-text-3
